### ISSUE TITLE
Use instance history caching to improve message processing throughput

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#82]
Currently we are slow to process orchestrator function messages because we fetch the full history from Azure Storage every time we process a batch of messages for that instance. Some internal MRU caching should help speed things up, especially for highly parallel fan-out/fan-in scenarios.

**Comment by SimonLuckenuik** [comment]
What about adding a snapshot concept to prevent rebuilding from the whole history? See this [article](https://msdn.microsoft.com/en-us/library/jj591577.aspx) for details, this is common in Event Sourcing scenarios.

**Comment by cgillum** [comment]
Yes, something like this would be better. I'm having some discussions with the Durable Task Framework maintainers to see if we can build that into the core framework so that all providers (including the Azure Storage provider used by Durable Functions) can participate. Thanks for the link.

**Comment by cgillum** [comment]
I added a new feature known as "extended sessions" to the Durable Task Framework to address this. It works by keeping the orchestration state in memory for a period of time to avoid replaying the orchestration when new messages continually arrive. The DurableTask.AzureStorage provider starting with version 1.2.0 will include these optimizations. We've merged in support for this in the latest Durable Functions code, and it will be available in the next release.
